internal/tool/tool.go

```diff
@@ -30,27 +39,195 @@ var _ codacy.Tool = (*codacySemgrep)(nil)
-	// }
+	configFile, err = createConfigFile(toolExecution, toolExecution.Patterns)
+	if err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll("tmp/")
+	defer configFile.Close()
 
-	semgrepCmd := semgrepCommand(configFile, *toolExecution.Files, toolExecution.SourceDir)
+	err = populateFilesByLanguage(toolExecution.Files, toolExecution.SourceDir)
+	if err != nil {
+		return nil, errors.New("Error getting files to analyse: " + err.Error())
+	}
 
-	semgrepOutput, semgrepError, err := runCommand(semgrepCmd)
+	result, err := run(configFile, toolExecution)
 	if err != nil {
-		return nil, errors.New("Error running semgrep: " + semgrepError)
+		return nil, err
+	}
+
+	return result, nil
+}
+
+func run(configFile *os.File, toolExecution codacy.ToolExecution) ([]codacy.Result, error) {
+	var result []codacy.Result
+	for language, files := range filesByLanguage {
+		semgrepCmd := semgrepCommand(configFile, toolExecution.SourceDir, language, files)
+
+		semgrepOutput, semgrepError, err := runCommand(semgrepCmd)
+		if err != nil {
+			return nil, errors.New("Error running semgrep: " + semgrepError)
+		}
+
+		output, err := parseOutput(toolExecution.ToolDefinition, semgrepOutput)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, output...)
 	}
 
-	result := parseOutput(toolExecution.ToolDefinition, semgrepOutput)
 	return result, nil
 }
 
+func populateFilesByLanguage(toolExecutionFiles *[]string, toolExecutionSourceDir string) error {
+	// If there are files to analyse, analyse only those files
+	if toolExecutionFiles != nil && len(*toolExecutionFiles) > 0 {
+		return populateFilesByLanguageFromFiles(*toolExecutionFiles)
+	}
+	// If there are no files to analyse, analyse all files from source dir
+	return populateFilesByLanguageFromSourceDir(toolExecutionSourceDir)
+}
+
+func populateFilesByLanguageFromFiles(toolExecutionFiles []string) error {
+	for _, file := range toolExecutionFiles {
+		addFileToFilesByLanguage(file)
+	}
+
+	return nil
+}
+
+func populateFilesByLanguageFromSourceDir(toolExecutionSourceDir string) error {
+	// Semgrep can analyse full directories and its subdirectories
+	// but we will have to analyse every extension from every file
+	// so we will have to do this walk somewhere else if we dont do it here
+	err := filepath.WalkDir(toolExecutionSourceDir, func(path string, info fs.DirEntry, walkErr error) error {
+		if walkErr != nil {
+			return walkErr
+		}
+		pathInfo, pathErr := info.Info()
+		if pathErr != nil {
+			return pathErr
+		}
+		// if it is a file and it is not a hidden file
+		if !pathInfo.IsDir() && !strings.HasPrefix(pathInfo.Name(), ".") {
+			addFileToFilesByLanguage(path)
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func addFileToFilesByLanguage(fileName string) {
+	language := detectLanguage(fileName)
+	filesByLanguage[language] = append(filesByLanguage[language], fileName)
+}
+
+// This feels illegal
```

Comment:
I have the same feeling about it for the converters I created for the docgenerator 😁 
perhaps having a const map{string, string} is better?